File: leetcode/cpp/P0358_RearrangeStringKDistApart.cpp
// https://leetcode.com/problems/rearrange-string-k-distance-apart/

#include <iostream>
#include <queue>
#include <string>
#include <unordered_map>

class Solution {
 public:
  std::string rearrangeString(std::string s, int k) {
    std::unordered_map<char, int> frequency;
    for (char ch : s) {
      frequency[ch] += 1;
    }

    std::unordered_map<char, int> prev_idx;
    std::priority_queue<std::pair<int, char>> pq;
    for (const auto& [ch, f] : frequency) {
      pq.push(std::make_pair(f, ch));
      prev_idx[ch] = -k;
    }

    std::string result = "";
    std::queue<std::pair<int, char>> buffer;
    for (int crt_idx = 0; crt_idx < s.length(); ++crt_idx) {
      bool ok = false;
      int count = -1;
      char ch = '*';
      while (pq.size() > 0) {
        std::pair<int, char> crt = pq.top();

        pq.pop();
        if (crt_idx - prev_idx[crt.second] >= k) {
          result += crt.second;
          prev_idx[crt.second] = crt_idx;
          ok = true;
          count = crt.first - 1;
          ch = crt.second;
          break;
        } else {
          buffer.push(crt);
        }
      }

      if (!ok) {
        return "";
      }

      if (count > 0) {
        pq.push(std::make_pair(count, ch));
      }
      // TODO: keep track of queue size.
      while (buffer.size() > 0) {
        pq.push(buffer.front());
        buffer.pop();
      }
    }

    return result;
  }
};

int main(int argc, char** argv) {
  // Note: solutions are not necessarily unique.
  Solution s;
  std::cout << "Example: aabbcc, 3 "
            << "Expected: "
            << "abcabc "
            << "Actual: " << s.rearrangeString("aabbcc", 3) << std::endl;
  return 0;
}

File: leetcode/python/P0358_RearrangeStringKDistApart.py
# https://leetcode.com/problems/rearrange-string-k-distance-apart/

from collections import Counter
import heapq


class Solution:
    def rearrangeString(self, s: str, k: int) -> str:
        # keep a priority queue, ordered by frequency;
        # if the current character can't be inserted, keep it in a buffer.
        # mark if can add the current character to the result, if not,
        # return empty string;
        # otherwise, add the character back to the queue, with updated
        # frequency, if we still have occurences.

        # from editorial - the buffer could be repalced by a queue.
        # Time complexity: O(n * k log(k)), where n is the length of s.
        # since k is 26 --> O(n)
        # building heap -- k log k, where k is the number of unique characters.
        # for each character in the string, we run heappop / heappush;
        # worst case, pop all elements of heap and add them back.
        freq = Counter(s)
        unique_chars = freq.keys()

        heap = []

        for ch, frequency in freq.items():
            heapq.heappush(heap, (-frequency, ch))

        result = ""
        prev_idx = {ch: -k for ch in unique_chars}
        count, ch = -1, "*"

        for crt_idx in range(len(s)):
            buffer = []
            ok = False
            while len(heap) > 0:
                count, ch = heapq.heappop(heap)
                if crt_idx - prev_idx[ch] >= k:
                    result += ch
                    prev_idx[ch] = crt_idx
                    count += 1
                    ok = True
                    break
                else:
                    # TODO: use deque here.
                    buffer.append((count, ch))

            if not ok:
                return ""

            if count < 0:
                heapq.heappush(heap, (count, ch))
            for x in buffer:
                heapq.heappush(heap, x)

        return result


if __name__ == "__main__":
    # NOTE: solution might not be unique.
    for testcase, expected in (
        (("aabbcc", 3), "abcabc"),
        (("aaabc", 2), "abaca"),
        (("aaabc", 3), ""),
        (("aaadbbcc", 2), "abacabcd"),
    ):
        actual = Solution().rearrangeString(*testcase)
        print(f"Testcase: {testcase}. Expected: {expected}. Actual: {actual}")
